### PR TITLE
add support for prune deep layer

### PR DESCRIPTION
**Pull request by fsx950223**
add support for prune deep layer

### CODE REVIEW

File: tensorflow_model_optimization/python/core/sparsity/keras/pruning_callbacks_test.py [hunk 1]
```diff
@@ -55,6 +55,45 @@ def _pruned_model_setup(self, custom_training_loop=False):
...
+
+    class TestModel(tf.keras.Model):
+      def __init__(self):
+        super().__init__()
+        self.test_layer = TestLayer()
+
+      def call(self, x):
+        return self.test_layer(x)
+
+    x_train = np.random.rand(self._BATCH_SIZE, 10)
```
Comment (alanchiao) [inline tensorflow_model_optimization/python/core/sparsity/keras/pruning_callbacks_test.py:78]: To make this unit test focus on the important aspects of what's being tested (the specific model and callback), could you try to reuse self._pruned_model_setup like the other unit tests? It hides away the compile/loss/optimizer, which don't matter too much.

You can move out the line "pruned_model = prune.prune_low_magnitude(
        keras_test_utils.build_simple_dense_model())" from `_pruned_model_setup` to help.

**Comment by alanchiao** [comment]
@liyunlu0618  could you review?

File: tensorflow_model_optimization/python/core/sparsity/keras/pruning_callbacks_test.py [hunk 2]
```diff
@@ -55,6 +55,45 @@ def _pruned_model_setup(self, custom_training_loop=False):
...
+    pruned_model.fit(
+        x_train,
+        y_train,
+        batch_size=self._BATCH_SIZE,
+        epochs=3,
+        callbacks=[
+            pruning_callbacks.UpdatePruningStep()
+        ])
+
+    self.assertEqual(2, pruned_model.layers[0]._layers[0].pruning_step)
```
Comment (alanchiao) [inline tensorflow_model_optimization/python/core/sparsity/keras/pruning_callbacks_test.py:95]: For background for this PR, what's the difference between _layers and layers?

In this example, pruned_model.layers[0] is the TestLayer object.
I guess the issue is that tf.keras.layers.Layer does not have a `layers` attribute and only a `_layers` attribute?

I'm asking since ._layers is a private attribute that's not intended to be used. I can also understand the use case where TestLayer doesn't need the training functions/utilities associated with a tf.keras.Model, which does have the `layers` attribute. 

@pavithrasv: could you comment on this? We have a callback (UpdatePruningStep) that updates a variable that we add via a wrapper to every layer in a model. It's not able to access that variable inside the layers inside a layer, which would motivate this usage of a private attribute.

Comment (alanchiao) [reply tensorflow_model_optimization/python/core/sparsity/keras/pruning_callbacks_test.py]: @pavithrasv: ping on the above. thank you.

Comment (pavithrasv) [reply tensorflow_model_optimization/python/core/sparsity/keras/pruning_callbacks_test.py]: Sorry about the delay. Missed the earlier msg. I don't fully remember the history behind the `layers` vs `_layers` attribute naming itself. But when you get the weights of a layers, by default it will return the weights of all the sub-layers as well. `_auto_track_sub_layers` property tracks this and unless it is explicitly turned off, layer.variables/weights should return all the sub layer weights.

If you want to get a list of sub layers including the layer itself, we have this private utility `_flatten_layers()`. `_layers` will contain Trackable containers in addition to sub-layers.

Comment (fsx950223) [reply tensorflow_model_optimization/python/core/sparsity/keras/pruning_callbacks_test.py]: So I should change to ```_flatten_layers(). _layers```?
I can't get the prune_step variable via ```.layers```

Comment (pavithrasv) [reply tensorflow_model_optimization/python/core/sparsity/keras/pruning_callbacks_test.py]: Ideally calling a layer should return the sublayer weights as well. Do you know why the prune_step variable is not returned? Between `._layers` and `._flatten_layers`, `._flatten_layers` is the correct one to use.

Comment (fsx950223) [reply tensorflow_model_optimization/python/core/sparsity/keras/pruning_callbacks_test.py]: > Ideally calling a layer should return the sublayer weights as well. Do you know why the prune_step variable is not returned? Between `._layers` and `._flatten_layers`, `._flatten_layers` is the correct one to use.

```tuples = [(var,self.step) for var in self.model.variables if var.name.split(':')[0].split('/')[-1] == 'pruning_step']```
The above code could get pruning steps too. Should I change like this?

**Comment by fsx950223** [comment]
cc @pavithrasv
